lib/web/desktop.go

```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
-				return
-			case evt := <-eventsC:
-				if evt == nil {
-					h.log.Debug("reached end of playback")
+		// Handle incoming playback actions.
+		go func() {
+			defer h.log.Debug("playback action-recieving goroutine returned")
+			defer ps.Close()
+
+			for {
+				if !ps.GetWsOpen() {
 					return
 				}
-				switch e := evt.(type) {
-				case *apievents.DesktopRecording:
-					if e.DelayMilliseconds > lastDelay {
-						time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
-						lastDelay = e.DelayMilliseconds
-					}
-					if _, err := ws.Write(e.Message); err != nil {
-						h.log.WithError(err).Error("failed to write TDP message over websocket")
+
+				action := playbackAction{}
+				// websocket.JSON.Receive will return even if there's no
+				// message waiting on the websocket, in which case action will
+				// just have it's default values (i.e. action.Action == "")
+				err := websocket.JSON.Receive(ws, &action)
+				if err != nil {
+					h.log.WithError(err).Error("error reading from websocket")
+					return
+				}
+				if action.Action != "" {
+					h.log.Debugf("recieved playback action: %+v", action)
+					if action.Action == actionToggle {
+						ps.TogglePlaying()
+					} else {
+						h.log.Errorf("received unknown action: %v", action.Action)
 						return
 					}
+				}
+			}
+		}()
+
+		// Stream session events back to browser.
+		go func() {
+			defer h.log.Debug("playback event-streaming goroutine returned")
+			defer ps.Close()
+
+			var lastDelay int64
+			eventsC, errC := ctx.clt.StreamSessionEvents(rCtx, session.ID(sID), 0)
+			for {
+				if !ps.GetWsOpen() {
+					return
+				}
+
+				if !ps.GetPlaying() {
+					continue
+				}
+
+				select {
+				case err := <-errC:
+					h.log.WithError(err).Errorf("streaming session %v", sID)
+					return
+				case evt := <-eventsC:
+					if evt == nil {
+						h.log.Debug("reached end of playback, restarting")
+						lastDelay = 0
+						// TODO: this causes the session to be re-downloaded, make it smarter by checking for the file first?
+						eventsC, errC = ctx.clt.StreamSessionEvents(rCtx, session.ID(sID), 0)
+						continue
+					}
+					switch e := evt.(type) {
+					case *apievents.DesktopRecording:
+						if e.DelayMilliseconds > lastDelay {
+							time.Sleep(time.Duration(e.DelayMilliseconds-lastDelay) * time.Millisecond)
+							lastDelay = e.DelayMilliseconds
+						}
+						if _, err := ws.Write(e.Message); err != nil {
+							h.log.WithError(err).Error("failed to write TDP message over websocket")
+							return
+						}
+					default:
+						h.log.Warnf("session %v contains unexpected event type %T", sID, evt)
+					}
 				default:
-					h.log.Warnf("session %v contains unexpected event type %T", sID, evt)
+					continue
 				}
 			}
+		}()
+
+		// Hang until the request context is cancelled or the websocket is closed by another goroutine.
+		for {
+			// Return if websocket is closed by another goroutine.
+			if !ps.GetWsOpen() {
```

Comment:
Same here. Lots of really tight loops that check a boolean as quickly as possible. Let's come up with a proper signaling mechanism to alert the playback routine of the current state.